ssntp/client.go

```diff
@@ -77,6 +77,7 @@ type Client struct {
 	closed    chan struct{}
 
 	frameWg              sync.WaitGroup
+	frameWgMutex         sync.Mutex
```

Comment:
I might be misunderstanding the code, but a channel doesn't look like it would work here.  It's possible that an arbitrary number of client.frameWg.Add(1) calls are made, each of which will have a corresponding client.frameWg.Done().  But the one waiter in Close() doesn't explicitly know how many to waitgroup Add(1)'s happened and how many to await.  I recall now in our Golang class being told you shouldn't do Add(1) on wait groups as it's a sign of a design flaw.  Instead you want to know how many children you have do an Add(N) once, and then one wait.

I think this points at the actual issue here.

If Close() hits its client.frameWg.Wait() _just_ ahead of one of the arbitrary number of handleSSNTPServer() runs doing a client.frameWg.Add(1), then the Close() will wait for one less frame handling completion than it should and it's possible the client.frameRoutinesChannel is closed ahead of its writers.  If that's the issue, my introduction of a lock doesn't truly address that without changing the lock/unlock placement.

In this case I think ssntpStatus needs an ssntpClosing ahead of ssntpClosed.  We'd enter that state in the Close() and then client.frameWg.Wait().  In the handleSSNTPServer() we'd want to bail out if client.status==ssntpClosing rather than client.frameWg.Add(1) and go client.processSSNTPFrame(&frame).

Or..there is already the closed state checked and set in Close().  Perhaps it just needs checked in the non-error path of the read returning inside handleSSNTPServer().

I'll test a couple options here and refactor.